README.md
# Homebridge Davis WeatherLink Live

This is a Homebridge plugin that allows you to integrate your Davis WeatherLink Live station API.

## Installation

You must have Homebridge already installed, then just install the plugin by running `npm install -g homebridge-davisweatherlinklive`

## How It Works

This will call your Davis WeatherLink Live station API.  The endpoint it is designed for is /v1/current_conditions.  It will look at the following JSON path for temperature and humidity.

* Temperature - `data.conditions[0].temp`
* Humidity - `data.conditions[0].hum`

The API will be called once at start up and polled periodically.  The results will be stored in memory, to prevent slowness when opening the Home app.

## Configuration

I have included an example config of the accessory in `example.config.json`.

### Required Accessory Options

* `url` - The URL for the Davis WeatherLink Live station API.  e.g. http://10.0.0.100/v1/current_conditions
* `name` - Friendly name for the sensor
* `manufacturer` - Manufacturer
* `model` - Model

### Optional Accessory Option

* `pollingIntervalSeconds` - Set this to the seconds value you want to periodically poll your Davis WeatherLink Live station API.  DEFAULT: 300
* `temperatureUnitOfMeasure` - Choose "C" or "F".  DEFAULT: C

RELEASE_NOTES.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2020-xx-xx

### Added

- New visual identity by [@tylerfortune8](https://github.com/tylerfortune8).

### Changed

- Start using "changelog" over "change log" since it's the common usage.

### Removed

- Section about "changelog" vs "CHANGELOG".

### Fixed

- Fix typos in recent README changes.

## [1.0.0] - 2019-08-04

### Added

- Initial version released.


DavisWeatherLinkLive.js
class DavisWeatherLinkLive {
	constructor(arg1, arg2) {

	}
}

DavisWeatherLinkLive.prototype = {
    httpRequest: function (url, body, method, callback) {
        request({
			url: url,
			body: body,
			method: method
		},
		function (error, response, body) {
			callback(error, response, body)
		})
    },

    getStateHumidity: function (callback) {
        callback(null, this._cachedData.humidity);
    },

    getStateTemperature: function (callback) {
        callback(null, this._cachedData.temperature);
    },

    getServices: function () {
        var services = [],
            informationService = new Service.AccessoryInformation();

        informationService
            .setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
            .setCharacteristic(Characteristic.Model, this.model)
            .setCharacteristic(Characteristic.SerialNumber, this.serial);
        services.push(informationService);

		this.temperatureService = new Service.TemperatureSensor(this.name);
        this.temperatureService
            .getCharacteristic(Characteristic.CurrentTemperature)
            .on("get", this.getStateTemperature.bind(this));
        services.push(this.temperatureService);

		this.humidityService = new Service.HumiditySensor(this.name);
		this.humidityService
			.getCharacteristic(Characteristic.CurrentRelativeHumidity)
			.setProps({minValue: 0, maxValue: 100})
			.on("get", this.getStateHumidity.bind(this));
		services.push(this.humidityService);

        return services;
	},

	getData: function (url) {
		this.httpRequest(url, "", "GET", function (error, response, responseBody) {
			var queue = function () {
				if (this._timeoutID > -1) {
					clearTimeout(this._timeoutID);
					this._timeoutID = -1;
				}
	
				this._timeoutID = setTimeout(function () {
					this._timeoutID = -1;
					this.getData(this.url);
				}.bind(this), this.pollingIntervalSeconds * 1000);
			}.bind(this);

            if (error) {
				this.log("Request to Davis API failed: %s", error.message);
				queue();
                return;
			}
			
			this.log("Request to Davis API succeeded: %s", responseBody);

			var jsonResponse = JSON.parse(responseBody);
			
			if (jsonResponse.data && (!jsonResponse.data.conditions || jsonResponse.data.conditions.length == 0)) {
				this.log("Response from Davis API doesn't contain expected result.");
				queue();
				return;
			}

			var weatherResponse = jsonResponse.data.conditions[0];

			this._cachedData = {
				"temperature": this.temperatureUnitOfMeasure == "C" ? this.convertFromFahrenheitToCelsius(weatherResponse.temp) : weatherResponse.temp,
				"humidity": Math.round(weatherResponse.hum)
			};
			this.log("Successfully got data.  Temp %s, hum %s", this._cachedData.temperature, this._cachedData.humidity);

			queue();
        }.bind(this));
	},

	convertFromFahrenheitToCelsius: function (f) { //MUST BE A NUMBER!
        return parseFloat(((f - 32) * (5 / 9)).toFixed(1));
    }
};

index.js
var Service, Characteristic;
var request = require("request");
import Logging from "Logging";

module.exports = function (homebridge) {
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    homebridge.registerAccessory("homebridge-davisweatherlinklive", "DavisWeatherLinkLive", DavisWeatherLinkLive);
}

function DavisWeatherLinkLive(log, config) {
    this.log = log;

    // Config
    this.url = config["url"];
    this.name = config["name"];
    this.manufacturer = config["manufacturer"] || "Davis";
	this.model = config["model"] || "Default";
	this.pollingIntervalSeconds = parseInt(config["pollingIntervalSeconds"] || 300);
	this.temperatureUnitOfMeasure = (config["temperatureUnitOfMeasure"] || "C").toUpperCase();
	this.logLevel = (config["logLevel"] || Logging.LOG_LEVEL.Info;
	this._timeoutID = -1;
	this._cachedData = { "temperature": 0, "humidity": 0 };

	this.logger = new this.logger(log, config.logLevel);
	this.getData(this.url);
}

DavisWeatherLinkLive.prototype = {
    httpRequest: function (url, body, method, callback) {
        request({
			url: url,
			body: body,
			method: method
		},
		function (error, response, body) {
			callback(error, response, body)
		})
    },

    getStateHumidity: function (callback) {
        callback(null, this._cachedData.humidity);
    },

    getStateTemperature: function (callback) {
        callback(null, this._cachedData.temperature);
    },

    getServices: function () {
        var services = [],
            informationService = new Service.AccessoryInformation();

        informationService
            .setCharacteristic(Characteristic.Manufacturer, this.manufacturer)
            .setCharacteristic(Characteristic.Model, this.model)
            .setCharacteristic(Characteristic.SerialNumber, this.serial);
        services.push(informationService);

		this.temperatureService = new Service.TemperatureSensor(this.name);
        this.temperatureService
            .getCharacteristic(Characteristic.CurrentTemperature)
            .on("get", this.getStateTemperature.bind(this));
        services.push(this.temperatureService);

		this.humidityService = new Service.HumiditySensor(this.name);
		this.humidityService
			.getCharacteristic(Characteristic.CurrentRelativeHumidity)
			.setProps({minValue: 0, maxValue: 100})
			.on("get", this.getStateHumidity.bind(this));
		services.push(this.humidityService);

        return services;
	},

	getData: function (url) {
		this.httpRequest(url, "", "GET", function (error, response, responseBody) {
			var queue = function () {
				if (this._timeoutID > -1) {
					clearTimeout(this._timeoutID);
					this._timeoutID = -1;
				}
	
				this._timeoutID = setTimeout(function () {
					this._timeoutID = -1;
					this.getData(this.url);
				}.bind(this), this.pollingIntervalSeconds * 1000);
			}.bind(this);

            if (error) {
				this.log("Request to Davis API failed: %s", error.message);
				queue();
                return;
			}
			
			this.log("Request to Davis API succeeded: %s", responseBody);

			var jsonResponse = JSON.parse(responseBody);
			
			if (jsonResponse.data && (!jsonResponse.data.conditions || jsonResponse.data.conditions.length == 0)) {
				this.log("Response from Davis API doesn't contain expected result.");
				queue();
				return;
			}

			var weatherResponse = jsonResponse.data.conditions[0];

			this._cachedData = {
				"temperature": this.temperatureUnitOfMeasure == "C" ? this.convertFromFahrenheitToCelsius(weatherResponse.temp) : weatherResponse.temp,
				"humidity": Math.round(weatherResponse.hum)
			};
			this.log("Successfully got data.  Temp %s, hum %s", this._cachedData.temperature, this._cachedData.humidity);

			queue();
        }.bind(this));
	},

	convertFromFahrenheitToCelsius: function (f) { //MUST BE A NUMBER!
        return parseFloat(((f - 32) * (5 / 9)).toFixed(1));
    }
};